wren-main/src/main/java/io/wren/main/wireprotocol/MessagePlanner.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.wren.main.wireprotocol;
+
+import io.wren.main.wireprotocol.message.Describe;
+import io.wren.main.wireprotocol.message.DescribePortalAndExecute;
+import io.wren.main.wireprotocol.message.Execute;
+import io.wren.main.wireprotocol.message.Plan;
+import io.wren.main.wireprotocol.message.SendResult;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+public class MessagePlanner
+{
+    private MessagePlanner() {}
+
+    public static List<Plan> plan(Queue<Plan> messages)
+    {
+        List<Plan> plans = new ArrayList<>(messages.size());
+        Map<String, IndexedDescribePortal> portalDescriptions = new HashMap<>();
+        int i = 0;
+        while (!(messages.isEmpty())) {
+            Plan message = messages.poll();
+            if (message instanceof Describe.DescribePortal) {
+                Describe.DescribePortal describePortal = (Describe.DescribePortal) message;
+                portalDescriptions.put(describePortal.getTargetName(), new IndexedDescribePortal(i, describePortal));
+            }
+            if (message instanceof Execute) {
+                Execute execute = (Execute) message;
+                IndexedDescribePortal indexedDescribePortal = portalDescriptions.get(execute.getPortalName());
+                if (indexedDescribePortal != null) {
+                    DescribePortalAndExecute describePortalAndExecute = new DescribePortalAndExecute(execute);
+                    plans.set(indexedDescribePortal.index, describePortalAndExecute);
+                    plans.add(new SendResult(describePortalAndExecute));
```

Comment:
Curious as, to why we need to split the message into two plans `DescribePortalAndExecute` and `SendResult`. Seem the `DescribePortalAndExecute` should include sending results for the naming `Execute`.

Reply:
Because we cannot guarantee that "Execute" is sent immediately after "Describe". If there're something existed between them, the ordering of message should be ensured.
For example, 
```
PARSE 1,
BIND 1,
DescribePortal 1,
Parse 2, 
Execute 1
Sync
```
The response should be
```
ParseComplete
BindComplete
RowDescription
ParseComplete
DataRow
...
CommandComplete
ReadyForQuery
```

If the DescribePortalAndExecute do `SendResult`, the `DataRow` will be in front of the second ParseComplete.

Reply:
How about renaming it to `PendingDescribePortal` or `Waiting` without `Execute`?

Reply:
How about `ExecuteAndSendRowDescription`? It doesn't describe portal but send description according to `Execute`.